tree/20_cfe_basics/flag_lib.cf

```diff
@@ -0,0 +1,175 @@
+             persistence => "1440",
+             if => "free";
+
+    any::
+      "pass3" expression => "pass2";
+      "pass2" expression => "pass1";
+      "pass1" expression => "any";
+
+  methods:
+      "${flag_unique.value}" usebundle => get_flag("${name}"),
+                             useresult => "current";
+
+
+  reports:
+    pass3::
+      "ok"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "free.flag_reserved_${cname}";
+
+      "failed"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "!free";
+
+    pass3.info::
+      "Reserved flag '${cname}'"
+                          comment => "${flag_unique.value}",
+                               if => "free";
+
+    pass3.error::
+      "Failed, flag '${cname}' was not free"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "!free";
+}
+
+bundle agent get_flag(name) {
+  vars:
+      "cname" string => canonify("${name}");
+      "flag_unique.value" string => eval("${flag_unique.value}+1", "math", "infix");
+
+  classes:
+      "is_set" expression => isvariable("ncf_flags.data[${cname}]");
+      "pass1" expression => "any";
+
+  methods:
+      "${flag_unique.value}" usebundle => read_flags;
+
+  reports:
+      "set"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "is_set|flag_reserved_${cname}";
+
+      "free"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "!is_set.!flag_reserved_${cname}";
+    info::
+      "Flag '${cname}' is set"
+                          comment => "${flag_unique.value}",
+                               if => "is_set";
+
+      "Flag '${cname}' is reserved"
+                          comment => "${flag_unique.value}",
+                               if => "!is_set.flag_reserved_${cname}";
+
+      "Flag '${cname}' is free"
+                          comment => "${flag_unique.value}",
+                               if => "!is_set.!flag_reserved_${cname}";
+}
+
+bundle agent set_flag(name) {
+  vars:
+      "cname" string => canonify("${name}");
+      "time" string => strftime("localtime", "%FT%T%z", "${sys.systime}");
+      "flag" data => '{ "${cname}": { "created": "${time}" } }';
+    pass2::
+      "output" data => mergedata("ncf_flags.data", "flag");
+
+  classes:
+      "pass3" expression => "pass2";
+      "pass2" expression => "pass1";
+      "pass1" expression => "any";
+
+  files:
+    pass3::
+      "${configuration.flag_file}"
+        create => "true",
+        template_method => "mustache",
+        edit_template => "${sys.workdir}/modules/templates/datastate.json.tpl",
```

Comment:
you are missing this file in the PR

Reply:
It is already used for jinja templating.